### ISSUE TITLE
[2022] An ensemble of VisNet, Transformer-M, and pretraining models for molecular property prediction in OGB Large-Scale Challenge @ NeurIPS 2022

### ISSUE THREAD

**Issue by mjrs33** [mjrs33/paper#63]
## 概要
分子の最適構造と仮想生成構造の潜在空間の差を対照学習で小さくし、予測時に構造が利用できない場合でも高い予測精度を実現。さらに、複数のモダリティを扱うモデルとアンサンブルし、OGB-LSC(@NeurIPS 2022)で2位の成績を実現。

### 論文リンク
https://arxiv.org/abs/2211.12791

### 著者/所属機関
Wang, Yusong and Li, Shaoning and Wang, Tong and Wang, Zun and He, Xinheng and Shao, Bin and Liu, Tie-Yan

### 投稿日付
2022/11/23

## 背景
NeurIPS 2022では分子のHOMO-LUMO gapの予測コンペティション（OGB-LSCのPCQM4Mv2データセット）が行われた。
このデータセットは学習データには最適化された三次元構造が提供されるが、validation, testデータにはそれが無い。そのため、三次元構造をうまく活用するのはもちろん、トポロジーと構造間のギャップを埋めることが重要なタスクである。

## 新規性
* Transformer-Mに対しViSNetのvector-scalar操作を取り入れたTransformer-M-ViSNetを提案。
* 最適化された構造から、生成された構造に潜在表現を蒸留したPretrained-3D-ViSNetを提案

## 事前知識
ViSNetは角度や二面角をそのまま計算するのではなく、ノード間のベクトル等の内積計算に置き換えることで、計算量を $O(N)$ に抑えた手法である。
ViSNet内で用いられるベクトルの次元は通常3（3次元空間）であるが、球面調和関数で $2l+1$ の次元に容易に拡張可能である。

Transformer-Mは異なる分子の表現（2D, 3Dグラフ）を同じバックボーンモデルを用い、交互にそれぞれのチャンネルを活性化させるモデル。2D, 3Dを同時に学習させることで、それぞれの重要な知識を保持しつつ分子を表現できる。2DはSMILESから得られたグラフや特徴量、3Dはconformerの構造情報など。

## 手法
モデルの全体像は以下
![2E2C7B41-852F-4759-ADBA-D259969CA9E4](https://user-images.githubusercontent.com/40561031/212504571-1ad15495-cf93-436f-b4aa-9a85f4ec6b6a.jpeg)

Transdormer-M-ViSNetには2Dグラフと、最適化された3Dグラフを入力にし、複数のモダリティを同時に学習させる。

Pretrained-3D-ViSNetでは最適化された構造と生成された構造とのcontrastive lossをとり、知識を蒸留させる。validationとtestには最適化された3D構造が無いので、生成された構造の特徴を最適構造と特徴を一致させ、最適構造が無くても予測できるようにするということ。

最後にそれぞれのモデルの出力をアンサンブルし、予測値を出力する。

### Transformer-M-ViSNet
ノード特徴量は以下のように修正

$$X^{(0)} = X + \psi^{2D} + \psi^{3D RGC Angle}$$

$X$ は原子番号のembedding, $\psi$ はそれぞれの情報から得られたembedding。
また、
$$\psi^{3D RGC Angle} = [W_{as}\vec{V}, W_{at}\vec{V}]$$
であり、 $\vec{V}$ はノード特徴と変位ベクトルのテンソル積から得られる。
$W$ は同変性を満たすためバイアス無しのLinear層になっている。

attention層は以下で計算される
$$A(X) = softmax(\frac{XW_Q(XW_K)^T}{\sqrt{d}} + \psi^{2D} + \psi^{3D Distance} + \psi^{3D RGC Dihedral})$$

$\psi^{3D Distance}$ はTransformer-Mのユークリッド距離の合計（よく分からないので論文読む）。

### Pretrained-3D-ViSNet
CLIPと同様、RDKitで生成した構造と最適化された構造は、分子の物性を表現する2つのモダリティであると考えることができる。
しかし生成された構造はノイズが多く、潜在空間を正確に表現するのに悪影響を与える恐れがある。
そのため、まず最適化構造を入力としたViSNet-1を学習した。次に別のViSNet-2を用いて生成構造の潜在表現を学習する。このとき、ViSNet-1は固定し、最適化構造のembeddingと生成構造のembeddingのlossが小さくなるようにViSNet-2を対照学習させる。

これにより、最適化構造が無くても、生成した構造を用いて、最適化構造のembedding（に近いもの）を得ることができる。
得られたembeddingを2層のMLPに流し、出力を得る。

### 学習
Transformer-M-ViSNetを異なるハイパーパラメータで20種類、Pretrained-3D-ViSNetは2種類学習させ、全22種類のモデルのアンサンブルを出力値のうち、中間の出力10種類のアンサンブルを最終的な出力値とした。

## 結果
MAEで0.0723 eVを達成。

## 参考文献
ViSNet：
Y. Wang, S. Li, X. He, M. Li, Z. Wang, N. Zheng, B. Shao, T. Wang, and T.-Y. Liu. Visnet: a scalable and accurate geometric deep learning potential for molecular dynamics simulation. arXiv preprint arXiv:2210.16518, 2022.

Transformer-M：
S. Luo, T. Chen, Y. Xu, S. Zheng, T.-Y. Liu, L. Wang, and D. He. One transformer can understand both 2d & 3d molecular data. arXiv preprint arXiv:2210.01765, 2022
